### DISCUSSION TITLE
External SPI and QSPI memories

### DISCUSSION THREAD

**Discussion by gabsi-w** [micropython/micropython#15534] [STM32 / Pyboard]
Hello, I am porting Micropython to a custom hardware based on an STM32F746NE chip, which due to size constraints, not all pins are available, like the USB for example. The board is programmed through a segger j-link using J-Link commander, where I write direct in the memory the firmware bin files that I compiled through WSL. The whole system works just fine, I was able to connect to micropython via PUTTY and get an REPL. The microcontroller has two external 64Mb memories connected to it, one to QSPI bank 1 and another to SPI6 with a random GPIO pin as the chip select. I am checking the present examples to see how the other boards implement it, but since I am kinda new with it I have many questions, like what is the best approach, for example.

So, does it make sense to use the QSPI as the internal memory? Would I need to use the mboot in this case? Will mboot work without the USB?

And about the SPI, as I understood this one can't be used as internal memory as the SPI doens not support the memory mapping. So, to use it as a external memory, how could I address it since the SPI region in the memory does not have 64 Mb of space?

Last, but not the least, how should I work with the bdev, boardinit, mpcoonfigboard (.h and .k), what must I include or define in order to it to work? 

Any help will be welcomed. Some exxamples that I am studying are: STM32F7/F769Disco, PyBoard v2 and Arduino Giga

Thank you!

**Comment by AndreoBotelho** [comment]
QSPI can be memory mapped but is is not required, you can use it as normal filesystem, SPI can not be mapped but can be mounted as a filesystem too with OS module, why use two flash if you have low pins available? As you are using F746NE, you have 512kb of internal flash including the two extenal flash and all of then can be used by micropython, read mboot.md on /mboot directory from mpy source with (few) instructions on how to setup you board firmware and some board definitions already have mboot configured for one or two external flash that you can copy

**Reply by gabsi-w** [reply]
Thank you for your reply! 

So, about the pins. I mean, the board has not that many pins available, but the microcontroller itself is hardwire to two 8MB flash memories, onte to the QSPI and one to the SPI6. 
I already put QSPI to work by defining the region in the ld file, then defining the pins in mpconfigboard.h so the qspi file can use it as mapped memory.
I am now having problems with the SPI. I am not so familiarized with bootloaders, is it possible to have mboot without having the USB connection?

**Reply by AndreoBotelho** [reply]
mboot can use USB DFU mode or I2C port

**Reply by gabsi-w** [reply]
Thanks for the explanation.

I actually could make both memories work, but separately. Is there a way of using both together, withous using it in mboot mode?
